### PR TITLE
Add minimal WF200 driver

### PR DESCRIPTION
**Pull request by rvoell-gcx**
Add basic support for Silabs WF200 Wi-Fi chips.  This driver currently implements station mode and scanning for Wi-Fi networks.

I used the `wio_terminal` board to make twister build the wifi sample using this driver, as its the only board in the current zephyr source providing an raspberry pi 40-pin connector. The eval boards from silabs have a silabs specific connector and this raspberry pi connector. I do not have a `wio_terminal` board to test with, so i do not know if this actually works in this combination.

This PR has been held as minimal as possible, to ease the mainlining process. Additional Features will be added in subsequent PRs, such as
- AP mode
- Statistics API
- Status API
- Powermanagement

This PR depends on <https://github.com/zephyrproject-rtos/hal_silabs/pull/15>, which adds silabs FullMAC driver and the firmwareblob needed by the hardware.

I already tried to mainline this in the past, but i think i tried too much at once. The old PR was #37525

Things needed before merging:

- [ ] Approval from TSC about including binary firmware blob in `west blobs`, see https://github.com/zephyrproject-rtos/zephyr/issues/62305
- [ ] Merging of vendor driver library into HAL, see https://github.com/zephyrproject-rtos/hal_silabs/pull/15
- [ ] Renaming driver to 'wfx' or 'wf200'

### CODE REVIEW

**Comment by carlescufi** [comment]
Hi, please use the `pull/<num>/head` format in the `revision` field in `west.yml`.
See: https://docs.zephyrproject.org/latest/develop/modules.html#submitting-changes-to-modules

**Comment by rvoell-gcx** [comment]
Oh, of course :)

**Comment by rvoell-gcx** [comment]
I fixed to compliance problems so far, but i'm unsure how to fix the line length problems:
```
----------------------------------------------------------------------------------------
Commit ceabeed50ada ("drivers: wifi: Add driver for SiLabs WFX200 series of WiFi chips")
----------------------------------------------------------------------------------------
ceabeed50ada4e4cf4d9b46cea00d69d7ce0ab4a:1475: WARNING:LONG_LINE: line length of 330 exceeds 100 columns
#1475: FILE: include/zephyr/dt-bindings/wifi/silabs_wf200.h:36:
+       "{b:{a:{a:4,b:0,c:0,d:0,e:A},b:{a:4,b:0,c:0,d:0,e:B},c:{a:4,b:0,c:0,d:0,e:C},d:{a:4,b:0,c:0,d:0,e:D},e:{a:4,b:0,c:0,d:0,e:E},f:{a:4,b:0,c:0,d:0,e:F},g:{a:4,b:0,c:0,d:0,e:G},h:{a:4,b:0,c:0,d:0,e:H},i:{a:4,b:0,c:0,d:0,e:I},j:{a:4,b:0,c:0,d:0,e:J},k:{a:4,b:0,c:0,d:0,e:K},l:{a:4,b:0,c:0,d:1,e:L},m:{a:4,b:0,c:0,d:1,e:M}}}", \

ceabeed50ada4e4cf4d9b46cea00d69d7ce0ab4a:1478: WARNING:LONG_LINE: line length of 201 exceeds 100 columns
#1478: FILE: include/zephyr/dt-bindings/wifi/silabs_wf200.h:39:
+       "{h:{e:0,a:50,b:0,d:0,c:[{a:1,b:[0,0,0,0,0,0]},{a:2,b:[0,0,0,0,0,0]},{a:[3,9],b:[0,0,0,0,0,0]},{a:A,b:[0,0,0,0,0,0]},{a:B,b:[0,0,0,0,0,0]},{a:[C,D],b:[0,0,0,0,0,0]},{a:E,b:[0,0,0,0,0,0]}]}}", \

ceabeed50ada4e4cf4d9b46cea00d69d7ce0ab4a:1491: WARNING:LONG_LINE: line length of 330 exceeds 100 columns
#1491: FILE: include/zephyr/dt-bindings/wifi/silabs_wf200.h:52:
+       "{b:{a:{a:4,b:0,c:0,d:0,e:A},b:{a:4,b:0,c:0,d:0,e:B},c:{a:4,b:0,c:0,d:0,e:C},d:{a:4,b:0,c:0,d:0,e:D},e:{a:4,b:0,c:0,d:0,e:E},f:{a:4,b:0,c:0,d:0,e:F},g:{a:4,b:0,c:0,d:0,e:G},h:{a:4,b:0,c:0,d:0,e:H},i:{a:4,b:0,c:0,d:0,e:I},j:{a:4,b:0,c:0,d:0,e:J},k:{a:4,b:0,c:0,d:0,e:K},l:{a:4,b:0,c:0,d:1,e:L},m:{a:4,b:0,c:0,d:1,e:M}}}", \

ceabeed50ada4e4cf4d9b46cea00d69d7ce0ab4a:1493: WARNING:LONG_LINE: line length of 201 exceeds 100 columns
#1493: FILE: include/zephyr/dt-bindings/wifi/silabs_wf200.h:54:
+       "{h:{e:0,a:50,b:0,d:0,c:[{a:1,b:[0,0,0,0,0,0]},{a:2,b:[0,0,0,0,0,0]},{a:[3,9],b:[0,0,0,0,0,0]},{a:A,b:[0,0,0,0,0,0]},{a:B,b:[0,0,0,0,0,0]},{a:[C,D],b:[0,0,0,0,0,0]},{a:E,b:[0,0,0,0,0,0]}]}}", \

ceabeed50ada4e4cf4d9b46cea00d69d7ce0ab4a:1506: WARNING:LONG_LINE: line length of 330 exceeds 100 columns
#1506: FILE: include/zephyr/dt-bindings/wifi/silabs_wf200.h:67:
+       "{b:{a:{a:4,b:0,c:0,d:0,e:A},b:{a:4,b:0,c:0,d:0,e:B},c:{a:4,b:0,c:0,d:0,e:C},d:{a:4,b:0,c:0,d:0,e:D},e:{a:4,b:0,c:0,d:0,e:E},f:{a:4,b:0,c:0,d:0,e:F},g:{a:4,b:0,c:0,d:0,e:G},h:{a:4,b:0,c:0,d:0,e:H},i:{a:4,b:0,c:0,d:0,e:I},j:{a:4,b:0,c:0,d:0,e:J},k:{a:4,b:0,c:0,d:0,e:K},l:{a:4,b:0,c:0,d:1,e:L},m:{a:4,b:0,c:0,d:1,e:M}}}", \

ceabeed50ada4e4cf4d9b46cea00d69d7ce0ab4a:1508: WARNING:LONG_LINE: line length of 201 exceeds 100 columns
#1508: FILE: include/zephyr/dt-bindings/wifi/silabs_wf200.h:69:
+       "{h:{e:0,a:50,b:0,d:0,c:[{a:1,b:[0,0,0,0,0,0]},{a:2,b:[0,0,0,0,0,0]},{a:[3,9],b:[0,0,0,0,0,0]},{a:A,b:[0,0,0,0,0,0]},{a:B,b:[0,0,0,0,0,0]},{a:[C,D],b:[0,0,0,0,0,0]},{a:E,b:[0,0,0,0,0,0]}]}}", \

ceabeed50ada4e4cf4d9b46cea00d69d7ce0ab4a total: 0 errors, 6 warnings, 1426 lines checked
```

These are part of the predefined "Platform Data Set" configuration data, which the firmware needs to be given. The defines are intended to be used in the device tree. If the strings are broken into smaller strings, the device tree parser can not correctly use them anymore.

File: drivers/wifi/wfx200/wfx200_events.c [hunk 1]
```diff
@@ -0,0 +1,89 @@
+/**
+ * Copyright (c) 2023 grandcentrix GmbH
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <sl_wfx.h>
+#undef BIT
```
Comment (gmarull) [inline drivers/wifi/wfx200/wfx200_events.c:8]: -1, this is not allowed, fix `sl_wfx.h` instead.

**Review by gmarull — CHANGES_REQUESTED** [review]

File: boards/shields/silabs_wifi_expansion_kit/Kconfig.defconfig [hunk 2]
```diff
@@ -0,0 +1,25 @@
+# Copyright (c) 2023 grandcentrix GmbH
+# SPDX-License-Identifier: Apache-2.0
+
+if SHIELD_SILABS_WIFI_EXPANSION_KIT_SLEXP8023A || SHIELD_SILABS_WIFI_EXPANSION_KIT_SLEXP8022A
+
+if NETWORKING
+
+config SPI
+	default y
```
Comment (gmarull) [inline boards/shields/silabs_wifi_expansion_kit/Kconfig.defconfig:9]: redundant

File: boards/shields/silabs_wifi_expansion_kit/silabs_wifi_expansion_kit_slexp8022a.overlay [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+        compatible = "silabs,wfx200";
+        reg = <0>;
+        label = "wifi0";
+        spi-max-frequency = <5200000>;
+
+        int-gpios = <&raspberrypi_header 20 GPIO_ACTIVE_HIGH>;
+        wake-gpios = <&raspberrypi_header 21 GPIO_ACTIVE_HIGH>;
+        reset-gpios = <&raspberrypi_header 22 GPIO_ACTIVE_LOW>;
+
+        pds = SILABS_WF200_PDS_WF200_EVAL;
```
Comment (gmarull) [inline boards/shields/silabs_wifi_expansion_kit/silabs_wifi_expansion_kit_slexp8022a.overlay:21]: what's going on with this property?

File: drivers/wifi/wfx200/wfx200_events.c (hunk 1 above)
Comment (rvoell-gcx) [reply drivers/wifi/wfx200/wfx200_events.c:8]: First, thanks for your review.
I wanted to keep sl_wfx.h as near as possible to the original. I'll fix that in the PR for the HAL

File: boards/shields/silabs_wifi_expansion_kit/Kconfig.defconfig (hunk 2 above)
Comment (rvoell-gcx) [reply boards/shields/silabs_wifi_expansion_kit/Kconfig.defconfig:9]: Hmm, should i add an `depends on SPI` to `config WIFI_WFX200` instead?

File: boards/shields/silabs_wifi_expansion_kit/silabs_wifi_expansion_kit_slexp8022a.overlay (hunk 3 above)
Comment (rvoell-gcx) [reply boards/shields/silabs_wifi_expansion_kit/silabs_wifi_expansion_kit_slexp8022a.overlay:21]: About that, i hoped documenting it in the device tree binding would be enough.
These are json-ish looking strings, which are sent during initialization to the WF200 chip. It configures quite a lot of properties, like which antenna port to use, how to configure its oscillator etc.. I really would like have these properties configurable in the device tree as individual properties, but its quite a bit of effort. I thought about generating these using macro magic from the devicetree or adding the PDS Template for the user to fill out and then generating the PDS with silabs pds_compress script. I don't like that property as its currently is, but i wanted to keep the PR small for now.

File: drivers/wifi/wfx200/wfx200_events.c (hunk 1 above)
Comment (rvoell-gcx) [reply drivers/wifi/wfx200/wfx200_events.c]: I fixed that.

**Comment by jerome-pouiller** [comment]
Hello @rvoell-gcx,

For information, `WF200` is the name of the chip while `wfx` is the name the driver. The idea behind was `wfx` could also handle future generations of the WFxxx series (wf300, etc...).

So, I believe this driver should be named `wf200` or `wfx` (or even wfx00), but `wfx200` is not a wise choice.

**Comment by rvoell-gcx** [comment]
That actually makes sense. I was really unsure about the naming. The module i have here, is a wfm200. The only other device i knew about, is the wf200 and the silabs driver was named wfx, so wfx200 seemed so cover it all :grinning: . I think i would rename the driver to wf200 then, as this driver currently only has support for the wf200 series.

**Comment by jerome-pouiller** [comment]
> There are only two hard things in Computer Science: cache invalidation and naming things.
> _Phil Karlton_

**Pull request closed without merging**